Incremental solving producing cprover status as it goes

thk123 opened the pull request:
Needs some tidy up but this prints output that is consistent with the
old version

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Print incremental-status when doing incremental solving

It is not possible to conclude successful verification before
the unwinding has finished, so this prints inconclusive until
the analysis is complete. However, if one property has been
falsified, then even though the analysis can continue to check
other properties, the verification will certainly fail.

Commit:
Add tests verifying the incremental status is output

Commit:
Tests for incremental status with multiple properties

The incremental status should report failure from the first unwinding
for which one property is falsified.

peterschrammel commented on src/goto-checker/single_loop_incremental_symex_checker.cpp:
```diff
@@ -47,6 +47,23 @@ single_loop_incremental_symex_checkert::single_loop_incremental_symex_checkert(
+    [](const std::pair<irep_idt, property_infot> &property) {
+      return property.second.status == property_statust::FAIL;
+    });
+  std::string status = any_failures ? "FAILURE" : "INCONCLUSIVE";
```
How does this behave in the success case?
Please check the output also on a `VERIFICATION SUCCESSFUL` test.

NlightNFotis replied:
@peterschrammel In the commit message this says that it's not possible to conclude that verification is successful before the unwinding is finished, so it doesn't matter?

NlightNFotis commented on regression/cbmc-incr-oneloop/multiple-asserts/test.desc:
```diff
@@ -0,0 +1,17 @@
+--
+^warning: ignoring
+--
+Ensure that with multiple properites, the incremental stauts
```
Can this be changed? 

* `properites` -> `properties`
* `stauts` -> `status`

NlightNFotis commented on regression/cbmc-incr-oneloop/multiple-asserts/test-no-cp.desc:
```diff
@@ -0,0 +1,19 @@
+switches to failure as soon as one property has been falsified
+but keeps unwinding till it has resolved them all
+
+Use `--no-propagation` to ensure that even the output is the same regardless of whether the asserts are checked via the solver or the constant propogator
```
`propogator` -> `propagator`

NlightNFotis commented on regression/cbmc-incr-oneloop/multiple-asserts/test-no-cp.desc:
```diff
@@ -0,0 +1,19 @@
+--
+^warning: ignoring
+--
+Ensure that with multiple properites, the incremental stauts
```
ditto

NlightNFotis reviewed (approved):
Minor faults, it's good to go once they are fixed as far as I am concerned.

Commit:
Tweak where we print the incremental status

We want to print the status after we've done any solving for that
unwind, but before we've done an extra step of symexing.

We should not print in the event the full equation has been generated,
as then we have a full solution so ins't incremental.

Ditto we don't need to print the status if we've found a failure, as
this function will be called again to do the next symex invocation if
there are further propreties to check

Commit:
Adapt tests for printing

We now print after all solving, and the last thing before the next
increment, therefore the printing checks are easier to check immediately
precedes the next unwinding, rather than checking appears after the
increment we are talking about.

Commit:
Add tests for when the program verifies

In this case, the status is inconclusive until completion, at which
point success will be reported by CBMC overall.

Commit:
Add test to check for no incremental status when no properties to check

thk123 commented:
@peterschrammel @NlightNFotis comments addressed -> it required a little tweaking to exactly where the printing went to get it working correctly but I'm happy with it now. 

The behaviour is:

In each step, before doing the next round of unwinding, but after solving the equation, we will print inconclusive if there no failed properties yet, otherwise we'll print failed. Once we've either got an answer to all properties, or finished unwinding, no incremental status will be reported, since the final status will be reported next.
